README.md
# Realistic-Colonies

Brought to you by The Blue Surfer @ Fadoli and Wamilou.
Do not claim this addons as yours.

## What is Realistic-Colonies ?

A Garry's Mod addon that allows you and your friend to create and manage your own colony of antlions and headcrabs/zombies.

## Credits

- Namelezz and Overv ( initial coding + idea ) 
- The Blue Surfer (AKA fadoli) and Wamilou ( adding new content / updating / porting to gmod13 ) 

## What do yo need ?

To play Realistic-Colonies, you need :

- Garry's mod
- Half-Life 2

## How to install this addon ?

You have 2 ways to download this addon :

- From Steam Workshop(auto-update and install)
- From Github(you need yourself to install Realistic colonies and update it)

### From Steam Workshop

1. Go on [Steam Workshop page](http://steamcommunity.com/sharedfiles/filedetails/?id=139477348)
2. Then press 'Subscribe' button
3. Now, Garry's mod will download and update our addon for you.
     

### From Github
- Download the files from here
- Put them in your garrysmod/addons/


changelog.md
# Changelog

## Todo

- genetic thing or class thing

## 2.3.1 ( 31/05/2022 )

### Fixed

- [Plants] Huge plant model are now properly working (sorry)

## 2.3.0 ( 29/05/2022 )

### Changed

- Visual changes
  - [Plants] Implemented Model variation
  - [Plants] Implemented Skin variation (for models where they are available)
  - [Plants] Added variation to plant sizes (have no other effects than model scale)
- [NPCs/Plants] Minimal size is now 0.2 instead of 0.1
- [NPCs] Non mature NPCs will now fear others npcs groups

### Fixed

- [Plants] Small_plants growing animation is now smoother
- [Meat] Fixed an issue related to coloration of some meats causing LUA errors (most likely API changes in GMOD)

## 2.2.1 ( Hotfix 29/06/16 )

### Fixed

- Fixing a bug with rc_printevents, and message displaying
Thanks to "Combine DvL ID:3"

## 2.2 ( Workshop upload 23/06/16 ) : Pushing the Revs update on the workshop

## 2.1 rev C ( 06/06/16 ) : The plant update

### Changed

- Small plants now grow ( and spawn melon relativly to their size )
- Plants spawn melon more often,
- Small plants and melons should grow quite more often.
- Plants now have a random +- 10% melon spawning rate
- Plants should spawn melon a bit faster on increased rc_speed
- Watermelon decay way faster ( so that there can't be an area with tons of melon )

### Fixed

- Plant should now properly react to rc_speed changes.
- Melon should now properly spawn plants on the ground.
- Small and Medium plant now have a correct growing timer
- Headcrab and Zombies behavior were incorrect when to hungry (not targeting some type of headcrab/zombie)

## 2.1 rev B (30/05/2016)

### Changed

- Npcs won't be able to lay egg as long as they are > 1/2 max hunger.
- Npcs will now start searching for food when they are a bit more hungry ( from 35% to 40% maxhunger )

### Balancing

- eggs restore more hunger
- meat restore more hunger
- minors melon hunger buffs

- Antlion lay less eggs ( in average )
- Headcrab are much less likely to be black headcrabs ( they can't kill their target )
- Headcrab are now a little bit likely to be fast headcrabs

- Plants Grow / die quite faster
- Smaller size plant are buffed ( spawn melon a bit more often and a little bit bigger )

- Baby now start with half of their max hitpoint
- Baby now start with 25% their hunger.

### Fixed

- Headcrab will now attack zombie when to hungry

### Known bug

- Human are not fighting back against headcrabs and zombies.

## 2.1 rev A (29/05/2016)

### Changed

- NPCs wont all target the same 'food', which improve the global efficiency of the colonies ( and as well using less time the costy searching for food function )

### Internal

- New function to init food items : rc_api.setFood ( ent , foodType[type : string] )
- New function to properly remove NPCs : rc_api.removeNPC ( ent )

## 2.1

### Changed

- Name in spawn tab changed
- Gamemode removed

### Fixed

- Npcs are not laying eggs when they are to hungry ( this really allowed colony to survive without having to eat )
- Fixed some mistakes in convar uses related to eggs
- Fixed some terrible mistakes in counting colonies member ( human was sometimes counting headcrab one instead ... )

## 2.0

- SPAWN ICONS ... yeah finally ... ( it took me a long time to make them =( )
Note : not all of them are included nor really good, but that's better than nothing

Plants :

- Changed some convar related things.
- Small plants spawn small melons, Big plants spawn big melon.
- Melon are spawn at higher places to prevent them geting stuck in the map ( you should not let tree grows in corner, they need a lot of space ! )

NPCS :

- Changed a lot of convar value to balance things
- Added a young state, npcs grow fast then stop ( young ) then grow again.
- NPCS are not eating their own meat, they won't target it, but they still become crazy if not enough food is present ( and very hungry , this does not affect human )
- NPCS can now eat eggs to stay alive
- Zombie drops 1 human and 1 headcrabs meat on death. (non natural death)
-Human now have crowbar !

Note :
Zombie are considered as headcrab, they can eat human meat but not headcrab one.
Zombie eggs are considered as human, they can be eat by other zombies and headcrab but not by human.

Things you will never notice :

- Changed a lot of internal code so the entities' script are much much smaller
- Improved the RC API.

## 2.0 pre release 5

(added) Torso zombies.

(Fix) Human death message are not showing as headcrab's ones.
(Fix) Headcrab are now attacking black headcrab when to hungry.

(Change) New addons format, so I can finally upload it to workshop ;).

## 2.0 pre release 4

- New trees stuff :
-- Randoming wich tree have to grow up
-- Increased melon spawning zone.
-- Tree limit increased
-- Tree are living older
-- Tree are spawning melon faster
-- Melon and tree are removing if the limit of the next tree is hit.
- Optimisation of food and trees,
- Bug fix related to rc_speed and rc_time (egg, meet, healing of npc)
- Melon now spawn on the ground.
- Npc can't lay egg when they are starving.
- some Balance

## 2.0 pre release 3

- The entities tab is now 'Realistic_Colonies' instead of 'Real_Colonies'.

- Adding human.
- Adding human egg.
- Adding human meat.
- Human can't eat human meat.

- Zombies egg look like human egg.
- Zombies are now droping human meat + headcrab meat instead of 3 headcrab meat.
- Zombies may now spawn from combat.
- Zombies now have there own hunger convar.

- Melon are now killable again (i'm trying a way so they explode when damaged.)

- Removed a gui thing.
- Optimized gui thing.
- Client side vars is now clientside xD

- Npcs may now eat human meat.
- Npcs may now eat others npc's egg.

Code side :

- Meat now have correct name.
- Plant now have correct name

## 2.0 pre release 2

- melon are not colony killer :).
- information writen over npc head are now related to their scale.
- fixed weapon to controll npc.
- fixed an other bug with weapon.

lua/autorun/rcolonies.lua
-- ======================--
-- =	Realistic      =--
-- =	Colonies       =--
-- ======================--
AddCSLuaFile("rcoloniesclient.lua")

coloniesnames = {"Bob", "Frank", "John", "David", "Mark", "Carl", "Russell", "Neil", "Amy", "Chris", "Aimiee", "Sarah",
                 "Kent", "Barry", "Scott", "Christine", "Jenny", "Josh", "Alien", "Scotty", "Fred", "Garry", "Erik",
                 "Harry", "Jean", "Harry", "Abby", "Alfred", "Amy", "Andy", "Abe", "Bart", "Bryan", "Bryce", "Alice",
                 "Ashley", "Gordon", "Mafiatoss", "Cake", "Jean", "Pipe", "Cupcake", "Mr Troll", "Kerrigan", "Jim",
                 "Raynor", "Douglas", "Adams", "Link", "Zelda", "Arthas", "Sharlock", "Homes", "Simpson", "Homer",
                 "Edisier", "Christophe", "Eric", "Hugo", "Victor", "Renner", "Denoel", "Poivre", "Sel", "Saarlouis",
                 "RitzenHoff", "Terraria", "Seb", "Pipi", "Caca", "Hint", "Bill"}

-- Global Convar
CreateConVar("rc_printevents", 1, {FCVAR_NOTIFY}) -- Shall we print death events ?
-- When set to 0 : not detail
-- When set to 1 or more : print starvation message
-- When set to 2 or more : print natural death message
CreateConVar("rc_hateplayers", 0, {FCVAR_NOTIFY})
CreateConVar("rc_spreadthelove", 0, {FCVAR_NOTIFY})
-- When set to 1, the colonies will try to kill player
CreateConVar("rc_searchrad", 2048, {FCVAR_NOTIFY})
-- Radius to search for food
CreateConVar("rc_remove", 0, {FCVAR_NOTIFY})
-- Can be used to cleanup all the addons entity

CreateConVar("rc_time", 1, {FCVAR_NOTIFY})
-- Delay in seconds between each execution of npcs stuff.
CreateConVar("rc_planttime", 1, {FCVAR_NOTIFY})
-- Delay in seconds between each execution of plant stuff.
CreateConVar("rc_speed", 1, {FCVAR_NOTIFY})
-- Changes the speed of the addons. (puting 2 will make everything 2 times faster)

-- Meat stuff
CreateConVar("rc_meat_time", 40, {FCVAR_NOTIFY}) -- Time needed for the meat to delete itself.

-- Watermelon stuff
CreateConVar("rc_watermelon_time", 20, {FCVAR_NOTIFY}) -- Time needed for a watermelon to become a plant.
CreateConVar("rc_watermelon_max", 100, {FCVAR_NOTIFY}) -- Limit of watermelon.

-- Food value
CreateConVar("rc_hungry", 40, {FCVAR_NOTIFY}) -- the % of maxhunger after which npcs will start searching for food
CreateConVar("rc_veryhungry", 80, {FCVAR_NOTIFY}) -- the % of maxhunger after which npcs of same class can eat each others (It doesn't affect human.)
CreateConVar("rc_food_melon", 45, {FCVAR_NOTIFY}) -- Food value of melon
CreateConVar("rc_food_antlion_meat", 50, {FCVAR_NOTIFY}) -- Food value of antlion's meat
CreateConVar("rc_food_headcrab_meat", 25, {FCVAR_NOTIFY}) -- Food value of headcrab's meat
CreateConVar("rc_food_human_meat", 50, {FCVAR_NOTIFY}) -- Food value of human's meat

CreateConVar("rc_food_antlion_egg", 40, {FCVAR_NOTIFY}) -- Food value of antlion's egg
CreateConVar("rc_food_headcrab_egg", 15, {FCVAR_NOTIFY}) -- Food value of headcrab's egg
CreateConVar("rc_food_human_egg", 30, {FCVAR_NOTIFY}) -- Food value of human's egg
CreateConVar("rc_food_zombie_egg", 30, {FCVAR_NOTIFY}) -- Food value of zombie's egg

-- Watermelon plant stuff
CreateConVar("rc_watermelonh_time", 7, {FCVAR_NOTIFY}) -- Time between each new watermelon
CreateConVar("rc_watermelonb_time", 9, {FCVAR_NOTIFY}) -- Time between each new watermelon
CreateConVar("rc_watermelonm_time", 12, {FCVAR_NOTIFY}) -- Time between each new watermelon
CreateConVar("rc_watermelons_time", 15, {FCVAR_NOTIFY}) -- Time between each new watermelon

CreateConVar("rc_watermelonh_size", 125, {FCVAR_NOTIFY}) -- Size of melon in %
CreateConVar("rc_watermelonb_size", 100, {FCVAR_NOTIFY}) -- Size of melon in %
CreateConVar("rc_watermelonm_size", 90, {FCVAR_NOTIFY}) -- Size of melon in %
CreateConVar("rc_watermelons_size", 75, {FCVAR_NOTIFY}) -- Size of melon in %

CreateConVar("rc_watermelonh_life", 900, {FCVAR_NOTIFY}) -- Time between a tree death
CreateConVar("rc_watermelonb_life", 400, {FCVAR_NOTIFY}) -- Time between tree grow
CreateConVar("rc_watermelonm_life", 120, {FCVAR_NOTIFY}) -- Time between tree grow
CreateConVar("rc_watermelons_life", 80, {FCVAR_NOTIFY}) -- Time between tree grow
CreateConVar("rc_tree_maxh", 3, {FCVAR_NOTIFY}) -- Maximum number of plant 
CreateConVar("rc_tree_maxb", 12, {FCVAR_NOTIFY}) -- Maximum number of plant 
CreateConVar("rc_tree_maxm", 20, {FCVAR_NOTIFY}) -- Maximum number of plant 
CreateConVar("rc_tree_maxs", 35, {FCVAR_NOTIFY}) -- Maximum number of plant 
CreateConVar("rc_watermelonh_distance", 900, {FCVAR_NOTIFY}) -- Maximum/Minimum distance watermelons can spawn from the plant
CreateConVar("rc_watermelonb_distance", 600, {FCVAR_NOTIFY}) -- Maximum/Minimum distance watermelons can spawn from the plant
CreateConVar("rc_watermelonm_distance", 300, {FCVAR_NOTIFY}) -- Maximum/Minimum distance watermelons can spawn from the plant
CreateConVar("rc_watermelons_distance", 150, {FCVAR_NOTIFY}) -- Maximum/Minimum distance watermelons can spawn from the plant

-- Antlion Vars
CreateConVar("rc_antlion_lifespan", 300, {FCVAR_NOTIFY}) -- Lifespan of antlion.
CreateConVar("rc_antlion_hunger", 1, {FCVAR_NOTIFY}) -- How fast should the hunger rise.
CreateConVar("rc_antlion_mhunger", 150, {FCVAR_NOTIFY}) -- The max hunger before npc die of starvation
CreateConVar("rc_antlion_eggtime", 10, {FCVAR_NOTIFY}) -- Time needed for the egg to break.
CreateConVar("rc_antlion_maturetime", 50, {FCVAR_NOTIFY}) -- Time needed for an antlion to become adult
CreateConVar("rc_antlion_pregtime", 80, {FCVAR_NOTIFY}) -- Time between each egg lay.
CreateConVar("rc_antlion_max", 50, {FCVAR_NOTIFY}) -- Maximum antlion population

-- Human Vars
CreateConVar("rc_human_lifespan", 450, {FCVAR_NOTIFY}) -- Lifespan of human.
CreateConVar("rc_human_hunger", 1, {FCVAR_NOTIFY}) -- How fast should the hunger rise.
CreateConVar("rc_human_mhunger", 200, {FCVAR_NOTIFY}) -- The max hunger before npc die of starvation
CreateConVar("rc_human_eggtime", 20, {FCVAR_NOTIFY}) -- Time needed for the egg to break.
CreateConVar("rc_human_maturetime", 80, {FCVAR_NOTIFY}) -- Time needed for an human to become adult
CreateConVar("rc_human_pregtime", 120, {FCVAR_NOTIFY}) -- Time between each egg lay.
CreateConVar("rc_human_max", 40, {FCVAR_NOTIFY}) -- Maximum human population

-- Headcrab Vars
CreateConVar("rc_zombie_lifespan", 400, {FCVAR_NOTIFY}) -- Lifespan of zombies
CreateConVar("rc_zombie_hunger", 1, {FCVAR_NOTIFY}) -- How quickly the hunger increases
CreateConVar("rc_zombie_max", 35, {FCVAR_NOTIFY}) -- Maximum zombie population
CreateConVar("rc_zombie_mhunger", 225, {FCVAR_NOTIFY}) -- Maximum hunger of zombies.
CreateConVar("rc_zombie_maturetime", 50, {FCVAR_NOTIFY}) -- How long before the zombie matures
CreateConVar("rc_zombie_pregtime", 75, {FCVAR_NOTIFY}) -- How long between each egg
CreateConVar("rc_zombie_eggtime", 15, {FCVAR_NOTIFY}) -- How long before egg hatches

CreateConVar("rc_headcrab_lifespan", 200, {FCVAR_NOTIFY}) -- How long the headcrab lives for
CreateConVar("rc_headcrab_hunger", 1, {FCVAR_NOTIFY}) -- How quickly the hunger increases
CreateConVar("rc_headcrab_mhunger", 115, {FCVAR_NOTIFY}) -- Maximum hunger of headcrab.
CreateConVar("rc_headcrab_eggtime", 5, {FCVAR_NOTIFY}) -- How long before egg hatches
CreateConVar("rc_headcrab_maturetime", 20, {FCVAR_NOTIFY}) -- How long before the Headcrab matures
CreateConVar("rc_headcrab_pregtime", 50, {FCVAR_NOTIFY}) -- How long between each egg
CreateConVar("rc_headcrab_max", 95, {FCVAR_NOTIFY}) -- Maximum headcrab population
CreateConVar("rc_headcrab_tozombie", 1, {FCVAR_NOTIFY}) -- Should Headcrabs spawn zombies when they kill citizens

hook.Add("Initialize", "ColonyTags", function()
    local tags = GetConVarString("sv_tags") or " "
    if string.find(tags, "colonies") == nil then
        RunConsoleCommand("sv_tags", tags .. ",Realistic Colonies Beta")
    end
end)

hook.Add("OnNPCKilled", "ColonyKilledNPC", function(v, k, w)
    local Ok = 0

    if v:GetClass() == "npc_citizen" or v:GetClass() == "npc_rebel" then
        Ok = 1
    end

    if k:GetClass():sub(1, 12) == "npc_headcrab" and k:GetNWBool("HC") and Ok == 1 and
        GetConVarNumber("rc_headcrab_tozombie") == 1 then
        -- Remove both the headcrab and the killed NPC and replace with zombie
        local zomb = ents.Create("colonies_zombie")
        zomb:SetPos(v:GetPos())
        zomb:Spawn()
        v:Remove()
        k:Remove()
    end

end)

hook.Add("PlayerInitialSpawn", "ColoniesDefaultPlayerFaction", function(ply, sid)
    ply:SetPData("HCfaction", "Default Faction")
end)
-- Function to set player faction
function ColoniesPlayerFaction(ply, cmnd, args)
    if args[1] then
        ply:SetPData("HCfaction", args[1])
    end
end
concommand.Add("rc_myfaction", ColoniesPlayerFaction)
